### PR TITLE
MF-1198 - Add errors package tests

### LINKED ISSUE #1198: Add errors package tests
**FEATURE REQUEST**

2. Describe the feature you are requesting, as well as the possible use case(s) for it.
Errors package from `/pkg` is missing unit tests.

3. Indicate the importance of this feature to you (must-have, should-have, nice-to-have).
This is a must-have.

### PR DESCRIPTION
**Pull request by darkodraskovic**
Add unit tests to `errors` package from `pkg/`.

Resolves #1198

### CODE REVIEW

**Comment by blokovi** [comment]
Did you consider to include test cases when error is `nil` (for any of these tests)?

File: pkg/errors/errors_test.go [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+func wrap(n int) error {
+	if n < 1 {
+		return errors.New(strconv.Itoa(n))
+	}
+	return errors.Wrap(errors.New(strconv.Itoa(n)), wrap(n-1))
+}
+
+func message(level int) string {
+	msg := strconv.Itoa(level)
+	for i := level - 1; i >= 0; i-- {
```
Comment (manuio) [inline pkg/errors/errors_test.go]: I would prefer `for i := 0; i < level; i++ {`

File: pkg/errors/errors_test.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+		errMsg := err.Error()
+		assert.Equal(t, msg, errMsg, fmt.Sprintf("%s: expected %s got %s\n", tc.desc, msg, errMsg))
+		contained := errors.New(strconv.Itoa(rand.Intn(tc.level)))
+		contains := errors.Contains(err, contained)
+		assert.Equal(t, true, contains, fmt.Sprintf("%s: expected %v to contain %v\n", tc.desc, err, contained))
+	}
+}
+
+func wrap(n int) error {
+	if n < 1 {
```
Comment (manuio) [inline pkg/errors/errors_test.go]: I would use n == 0

File: pkg/errors/errors_test.go (hunk 1 above)
Comment (darkodraskovic) [reply pkg/errors/errors_test.go]: The order is not relevant. However, semantically, we expect that higher number errors wrap lower number errors. This function corresponds to the `wrap` function, i.e. produces an error message of a n level wrapped error.

File: pkg/errors/errors_test.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+		},
+		{
+			desc: "level 2 wrapped error",
+			err:  errors.Wrap(err1, errors.Wrap(err2, err3)),
+			msg:  fmt.Sprintf("%s : %s : %s", msg1, msg2, msg3),
+		},
+		{
+			desc: fmt.Sprintf("level %d wrapped error", level),
+			err:  wrap(level),
+			msg:  message(level),
```
Comment (manuio) [inline pkg/errors/errors_test.go:48]: Why you are not using error messages here?

**Review by manuio — CHANGES_REQUESTED** [review]

File: pkg/errors/errors_test.go (hunk 3 above)
Comment (darkodraskovic) [reply pkg/errors/errors_test.go:48]: Function `message` produces error message of n level wrapped message. We use this error message to compare it with the error of an actually wrapped error with func `err`.

File: pkg/errors/errors_test.go (hunk 1 above)
Comment (manuio) [reply pkg/errors/errors_test.go]: This is more readable: 
```
for i := 0; i < level; i++ {
    msg = fmt.Sprintf("%s : %s", msg, strconv.Itoa(level - i))
}
```

File: pkg/errors/errors_test.go (hunk 2 above)
Comment (darkodraskovic) [reply pkg/errors/errors_test.go]: Although in this case it is not possible, it is recommended to test for less then if the exact number is not important. For example, if you decrement twice, you'll be at the `n == -1` and that'll cause stack overflow. However, in your solution intention is more clear.

**Comment by darkodraskovic** [comment]
> Did you consider to include test cases when error is `nil` (for any of these tests)?
Actually, a very good idea :D and an omission from my side...

File: pkg/errors/errors_test.go (hunk 1 above)
Comment (darkodraskovic) [reply pkg/errors/errors_test.go]: This part is tricky for me `strconv.Itoa(level - i)`. Solution is nice, however... However, it does not work out of the  box.

Comment (darkodraskovic) [reply pkg/errors/errors_test.go]: You need this in order to work `msg = fmt.Sprintf("%s : %s", msg, strconv.Itoa(level - i - 1))`. It's kind of ugly.

File: pkg/errors/errors_test.go (hunk 3 above)
Comment (manuio) [reply pkg/errors/errors_test.go:48]: What I mean is that for other tests you have an error message. You could have a funcrtion `createWrap(msg, level)` that you could use to generate wrapped error in all your tests. For example, you could replace `fmt.Sprintf("%s : %s : %s", msg1, msg2, msg3)`  by `createWrap(msg, 3)`

File: pkg/errors/errors_test.go (hunk 1 above)
Comment (manuio) [reply pkg/errors/errors_test.go]: Start by 1 then:
```
for i := 1; i < level; i++ {
    msg = fmt.Sprintf("%s : %s", msg, strconv.Itoa(level - i))
}
```

Comment (darkodraskovic) [reply pkg/errors/errors_test.go]: It does not work. You need to do it like this: `msg = fmt.Sprintf("%s : %s", msg, strconv.Itoa(level - i - 1))`. However, maybe a sign of a poor design :)

File: pkg/errors/errors_test.go (hunk 3 above)
Comment (darkodraskovic) [reply pkg/errors/errors_test.go:48]: Yeah, maybe a unified func would be better. For now, I've replaced `fmt.Sprintf("%s : %s : %s", msg1, msg2, msg3)` with `message` func.

File: pkg/errors/errors_test.go (hunk 1 above)
Comment (manuio) [reply pkg/errors/errors_test.go]: I'm not sure then. But the inversed for loop can be confusing

File: pkg/errors/errors_test.go [hunk 4]
```diff
@@ -67,31 +59,66 @@ func TestContains(t *testing.T) {
 		desc      string
 		container error
 		contained error
+		contains  bool
 	}{
+		{
+			desc:      "nil contains nil",
+			container: nil,
+			contained: nil,
+			contains:  true,
```
Comment (manuio) [inline pkg/errors/errors_test.go:69]: I would use `err`, `contained` and `res` or something easier to differentiate

Comment (darkodraskovic) [reply pkg/errors/errors_test.go:69]: IMO, this is a nice triplet. You know instantly which error contains, which error is contained and what should be result.

Comment (darkodraskovic) [reply pkg/errors/errors_test.go:69]: @manuio I also use this "terminology" in `TestWrap` and also a similar double term like wrapper and wrapped. IMO, it's very clear what this all means, while `err` and `res` are too generic.

**Review by blokovi — APPROVED** [review]

**Review by manuio — APPROVED** [review]
LGTM
